### ISSUE TITLE
Does `CallbackArg` still make sense?

### ISSUE THREAD

**Issue by TheEdward162** [bytecodealliance/javy#394]
Hi.

As far as I can tell `CallbackArg` was added in https://github.com/bytecodealliance/javy/pull/309 so that there would be a lifetime on `wrap_callback` this and args parameters and they wouldn't have to get eagerly resolved to `JSValue`. However, a week later https://github.com/bytecodealliance/javy/pull/335 introduced the same lifetime on `JSValueRef`, which leads me to believe the `CallbackArg` struct is obsolete now.

This is painful to deal with because I have some existing functions which I hook into callbacks from JS similar to this:
```rs
fn higher_level_callback<H: MyStateTrait>(
    state: &mut H,
    context: &Context,
    _this: JsValue,
    args: &[JsValue],
) -> Result<JsValue, JSError> {
    let response = match args.get(0) {
         None => return Err(JSError::Type("some error".to_string())),
         Some(name) => format!("Hello {}", name)
    }

    Ok(context.value_from_str(&response).unwrap())
}

// then somewhere in a macro I set up the callback and set it on an object, which I later inject into the global context
{
    let state = $state.clone();
    let fun = $context.wrap_callback(move |context, this, args| {
        let result = $fn_impl(state.borrow_mut().deref_mut(), context, this, args).map_err(anyhow::Error::from); // this should also not be anyhow, but that's another issue
        // some logging
        result
    })?;
    $parent.set_property($key, fun)?;
}
```

Converting this to a newer approach leads to this:
```rs
fn __export_message_exchange<'ctx, H: MyStateTrait>(
    state: &mut H,
    _this: JSValueRef<'ctx>,
    args: &[JSValueRef<'ctx>],
) -> Result<JSValue, JSError> {
    let response = match args.get(0) {
         None => return Err(JSError::Type("some error".to_string())),
         Some(name) => format!("Hello {}", name)
    }

    Ok(JSValue::String(response))
}

// the macro
{
    let state = $state.clone();
    let fun = $context.wrap_callback(move |context, this, args| {
        let this = unsafe { this.inner_value() }; // this isn't unsaf, and is just a noop anyway
        let args: Vec<_> = unsafe { args.into_iter().map(|a| a.inner_value()).collect() }; // this map->collect is pointless
        let result = $fn_impl(state.borrow_mut().deref_mut(), this, args).map_err(anyhow::Error::from);
        // some logging
        result
    })?;
    $parent.set_property($key, fun)?;
}
```
I have to unwrap and collect all items of `args` to turn them into `JSValueRef<'ctx>` for no reason.

And so I believe the `CallbackArg` struct is no longer needed. If this is true and I'm not missing anything I can open a PR to remove it.

**Comment by saulecabrera** [comment]
> I would have to eagerly convert in the macro, which defeats the entire purpose of lazy evaluation

I might be missing something here, because even if you convert in the macro definition, you wouldn't pay for that conversion until the callback is invoked right?

---

I think generally I understand what your issues are, thanks for adding more details. To be fair, we hadn't had the need to add more features to `JSValue` in such a way that it provides parity with `JSValueRef<'_>` but that's certainly something I'd personally prefer to do. On the other hand, as a stop gap measure, I think that removing the `CallbackArg` makes sense for now. So if you're willing to do a PR that's something that we can certainly land.

**Comment by TheEdward162** [comment]
> I might be missing something here, because even if you convert in the macro definition, you wouldn't pay for that conversion until the callback is invoked right?

Yes, of course, only after it is invoked. In general though, I would like to not have to pay for the conversion at all unless I truly need it (which is decided by the `__export_message_exchange` function in my example, not by the macro. Hence I would like to pass to the function the cheapest option, but in case of `CallbackArg` I still have to pay for the unwrapping and `collect`). Of course it's not the worst thing ever, but I think it increases friction so I'd like to avoid that if possible.

I'll open the PR asap.

**Comment by saulecabrera** [comment]
I'm going to go ahead and close this one since https://github.com/bytecodealliance/javy/pull/395 was merged.
